Userland/Libraries/LibGUI/BoxLayout.cpp

```diff
@@ -100,102 +145,215 @@ void BoxLayout::run(Widget& widget)
+
+            if (overcommitment_if_all_same_min_size > 0 && total_growth_if_not_overcommitet > 0) {
+                extra_needed_space -= (overcommitment_if_all_same_min_size * extra_needed_space + (total_growth_if_not_overcommitet - 1)) / (total_growth_if_not_overcommitet);
+            }
 
-            // If the slice was more than we needed, return remained to available_size.
-            int remainder_to_give_back = item_size_with_full_slice - item.size;
-            available_size += remainder_to_give_back;
+            VERIFY(extra_needed_space >= 0);
+            VERIFY(uncommitted_size >= extra_needed_space);
 
-            if (item.max_size >= 0 && item.size == item.max_size) {
-                // We've hit the item's max size. Don't give it any more space.
+            item.size += extra_needed_space;
+            if (item.max_size.is_regular_value() && item.size > item.max_size.value_verify_regular())
+                item.size = item.max_size.value_verify_regular();
+            uncommitted_size -= item.size - item.min_size.value_or_zero_if_shrink_with_verify();
+        }
+    }
+
+    loop_counter = 0;
+    // Pass 5: Determine the size for the opportunistically growing items.
+    while (opportunistic_growth_item_count > 0 && uncommitted_size > 0) {
+        VERIFY(loop_counter++ < 200);
+        int opportunistic_growth_item_extra_size = uncommitted_size / opportunistic_growth_item_count;
+        int pixels = uncommitted_size - opportunistic_growth_item_count * opportunistic_growth_item_extra_size;
+        VERIFY(pixels >= 0);
+        for (auto& item : items) {
+            if (item.preferred_size != GUI::SpecialDimension::OpportunisticGrow || item.final || !item.widget)
+                continue;
+
+            int pixel = (pixels > 0 ? 1 : 0);
+            pixels -= pixel;
+            int previous_size = item.size;
+            item.size += opportunistic_growth_item_extra_size + pixel;
+            if (item.max_size.is_regular_value() && item.size >= item.max_size.value_verify_regular()) {
+                item.size = item.max_size.value_verify_regular();
                 item.final = true;
-                --unfinished_items;
+                opportunistic_growth_item_count--;
             }
+            uncommitted_size -= item.size - previous_size;
         }
     }
 
-    // Pass 3: Place the widgets.
+    // Pass 6: Place the widgets.
     int current_x = margins().left() + content_rect.x();
     int current_y = margins().top() + content_rect.y();
 
-    auto widget_rect_with_margins_subtracted = content_rect;
-    widget_rect_with_margins_subtracted.take_from_left(margins().left());
-    widget_rect_with_margins_subtracted.take_from_top(margins().top());
-    widget_rect_with_margins_subtracted.take_from_right(margins().right());
-    widget_rect_with_margins_subtracted.take_from_bottom(margins().bottom());
+    auto widget_rect_with_margins_subtracted = margins().applied_to(content_rect);
```

Comment:
This seems easy to split into one commit / a commit with a few small changes like this, too much stuff in the one already!